core/src/avm2/globals/flash/display/shader_job.rs

```diff
@@ -134,27 +136,65 @@ pub fn get_shader_args<'gc>(
                                 bitmap.bitmap_handle(activation.gc(), activation.context.renderer),
                             )
                         } else if let Some(byte_array) = input.as_bytearray() {
-                            let expected_len = (width * height * input_channels) as usize
-                                * std::mem::size_of::<f32>();
-                            assert_eq!(byte_array.len(), expected_len);
                             assert_eq!(byte_array.endian(), Endian::Little);
-                            ImageInputTexture::Bytes {
+
+                            let num_pixels = (width * height) as usize;
+
+                            let (bytes, _trailing) = byte_array.bytes().as_chunks::<4>();
+                            let floats = bytemuck::cast_slice::<[u8; 4], f32>(bytes);
+
+                            let err = || make_error_2165(activation, name);
+                            let data = match input_channels {
+                                1 => FloatPixelData::R(
+                                    collect_floats::<1>(floats, num_pixels).ok_or_else(err)?,
+                                ),
+                                2 => FloatPixelData::Rg(
+                                    collect_floats::<2>(floats, num_pixels).ok_or_else(err)?,
+                                ),
+                                3 => FloatPixelData::Rgb(
+                                    collect_floats::<3>(floats, num_pixels).ok_or_else(err)?,
+                                ),
+                                4 => FloatPixelData::Rgba(
+                                    collect_floats::<4>(floats, num_pixels).ok_or_else(err)?,
+                                ),
+                                _ => panic!("Unexpected number of channels: {input_channels}"),
+                            };
+
+                            ImageInputTexture::Floats {
                                 width,
                                 height,
-                                channels: input_channels,
-                                bytes: byte_array.read_at(0, byte_array.len()).unwrap().to_vec(),
+                                data,
                             }
                         } else if let Some(vector) = input.as_vector_storage() {
```

Comment:
Can't we somehow merge this branch with bytearray's one? (i.e. remap one to the other)